p4-16/psa/PSA.mdk

```diff
@@ -2249,6 +2249,64 @@ The control plane should provide a
 
 ```
 
+
+# Atomicity of control plane API operations
+
+All table add, delete, and modify operations must be atomic relative
+to packet forwarding.  That is, for every table `apply` operation, and
+every control plane operation on a table that adds, deletes, or
+modifies one table entry, the `apply` operation should behave as if
+that control plane operation has not yet occurred, or as if the
+control plane operation is complete.  The P4 program should never
+behave as if the control plane operation is partially complete.
+
+For example, suppose a table `T` has an action `A` with 100 total bits
+of action parameters, and the control plane has added a table entry
+with a search key `K` and action `A`.  Later the control plane
+performs an update operation on the entry with key `K` that leaves the
+key `K` the same, but changes the 100 bits of action parameters.
+Every packet doing an `apply` on table `T` and matching the entry with
+key `K` should execute action `A` with either the old 100 bits of
+action parameters, or the new 100 bits of action parameters.
+
+The P4 Runtime API may define individual API calls that can perform
+more than one single operation, as defined here.  If so, a PSA
+implementation need only ensure that each single operation is atomic.
+There is no requirement that a sequence of _multiple_ table entry add,
+delete, or update operations should be atomic.
+
+The same applies for all control plane API operations on externs,
+unless the control plane operation explicitly documents otherwise.
+
+In particular, ActionProfile and ActionSelector single operations,
+such as adding a member to a group, removing a member from a group,
+adding an empty group, deleting and empty group, or modifying the
+action parameters of an action added earlier to a group, should all be
+atomic.
+
+Also, a control plane read, or write, of a single element of a
+Register array should be atomic, and behave as if it occurred before
+or after (but not during) any P4 program's section of code labeled
+with the `@atomic` annotation.  There is _no_ control plane operation
+on a Register that can atomically read, then modify the value, then
+write the modified value.  If you desire such a capability, you should
```

Comment:
I would split this last sentence in the paragraph into a separate note (programmer note?).